airflow/models/dataset.py

```diff
@@ -100,6 +98,55 @@ def __hash__(self):
     def __repr__(self):
         return f"{self.__class__.__name__}(uri={self.uri!r}, extra={self.extra!r})"
 
+    @property
+    def canonical_uri(self):
+        """
+        Resolve the canonical uri for a dataset.
+
+        If the uri doesn't have an `airflow` scheme, return it as-is.
```

Comment:
just want to point out, this will leave a bit of a gap concerning canonicalization... given that for some services there is more than one way to represent a given resource.  not to mention the ambiguity in the spec itself.

what to do about it?  well, we could lean on the "experimental" aspect of this and change it later if we decide to implement fuller canonicalization.  alternatively, we could be a little more restrictive, and instead of adding this property (which advertises canonicalization) we could add a method like "render_airflow_uri" or something...